daemon/logger/jsonfilelog/read.go

```diff
@@ -121,96 +119,106 @@ func tailFile(f io.ReadSeeker, logWatcher *logger.LogWatcher, tail int, since ti
 	}
 }
 
-func followLogs(f *os.File, logWatcher *logger.LogWatcher, notifyRotate chan interface{}, since time.Time) {
-	dec := json.NewDecoder(f)
-	l := &jsonlog.JSONLog{}
+func (l *JSONFileLogger) followLogs(f *os.File, logWatcher *logger.LogWatcher, notifyRotate chan interface{}, since time.Time) {
+	var (
+		msg        *logger.Message
+		err        error
+		rotated    bool
+		readIndex  uint64
+		writeIndex uint64
+
+		dec        = json.NewDecoder(f)
+		log        = &jsonlog.JSONLog{}
+		subscriber = l.writeNotifier.Subscribe()
+		watchClose = logWatcher.WatchClose()
+	)
+
+	go func() {
+		for {
+			if _, ok := <-subscriber; !ok {
+				return
+			}
+			atomic.AddUint64(&writeIndex, 1)
+		}
+	}()
 
-	fileWatcher, err := filenotify.New()
-	if err != nil {
-		logWatcher.Err <- err
+	reopenLogFile := func() error {
+		f.Close()
+		f, err = os.Open(f.Name())
+		if err != nil {
+			return err
+		}
+		dec = json.NewDecoder(f)
+		rotated = true
+		return nil
 	}
-	defer fileWatcher.Close()
 
-	var retries int
-	for {
-		msg, err := decodeLogLine(dec, l)
-		if err != nil {
-			if err != io.EOF {
-				// try again because this shouldn't happen
-				if _, ok := err.(*json.SyntaxError); ok && retries <= maxJSONDecodeRetry {
-					dec = json.NewDecoder(f)
-					retries++
-					continue
-				}
+	readToEnd := func() {
+		for {
+			msg, err := decodeLogLine(dec, log)
+			if err != nil {
+				return
+			}
+			if !since.IsZero() && msg.Timestamp.Before(since) {
+				continue
+			}
+			logWatcher.Msg <- msg
+		}
+	}
 
-				// io.ErrUnexpectedEOF is returned from json.Decoder when there is
-				// remaining data in the parser's buffer while an io.EOF occurs.
-				// If the json logger writes a partial json log entry to the disk
-				// while at the same time the decoder tries to decode it, the race condition happens.
-				if err == io.ErrUnexpectedEOF && retries <= maxJSONDecodeRetry {
-					reader := io.MultiReader(dec.Buffered(), f)
-					dec = json.NewDecoder(reader)
-					retries++
-					continue
-				}
+	defer func() {
+		l.writeNotifier.Evict(subscriber)
+		if rotated {
```

Comment:
@cpuguy83 If the log file is not rotated it'll be closed at https://github.com/docker/docker/blob/master/daemon/logger/jsonfilelog/read.go#L66